### ISSUE TITLE
Linux & PortAudio: Can you get it to work with the ALSA default device (without PulseAudio)

### ISSUE THREAD

**Issue by hzulla** [supercollider/supercollider#1944]
Currently, I'm testing the PortAudio driver of scsynth on Linux.

The cmake command to prepare my build is:

`cmake -DCMAKE_BUILD_TYPE=Release -DAUDIOAPI=portaudio ..`

My machine is running XUbuntu 15.10, which comes with PulseAudio installed by default.

PortAudio is talking to PulseAudio using its ALSA compatibility layer. As described in #1943, the PulseAudio `default` ALSA device reports 32 in and out channels.

With this setup, the following works fine with sclang:

```
s.options.device="default"
s.boot
( { f = LFSaw.kr(0.4, 0, 24, LFSaw.kr([8,7.23], 0, 3, 80)).midicps; CombN.ar(SinOsc.ar(f, 0, 0.04), 0.2, 0.2, 4) }.play)
```

Ok, fine.

Now  [turn off PulseAudio completely](http://askubuntu.com/a/394872/274072) and try again. Now, using pure ALSA, the `default` ALSA device reports 128 in and out channels and the previous sclang test code will result in very choppy audio on my machine.

Can others confirm this problem on their setups?

**Comment by timblechmann** [comment]
in my experience the quality of portaudio's backends differs a lot. especially the quality of the linux backends used to be rather bad, which was the reason, why supernova has a native jack backend. my recommendation is to take it with a grain of salt

**Comment by hzulla** [comment]
@timblechmann yes, seems like it. I'm investigating this and if it can't be fixed easily, an AUDIOAPI alternative for ALSA might be the better solution.

**Comment by hzulla** [comment]
Ok, using [a suggested latency value of 1.0s](https://lists.columbia.edu/pipermail/portaudio/2016-April/000376.html) seems to fix this on pure ALSA, but frankly that doesn't look like a good fix.

**Comment by hzulla** [comment]
Ok, another fix is to use no input. Audio playback is smooth once I put

`mDeviceInOut[0] = paNoDevice;`

in [line 347 here](https://github.com/hzulla/supercollider/blob/audiodriver/server/scsynth/SC_PortAudio.cpp#L347).

I'm guessing here, but probably it's the audio processing loop in the callback?

Also I have to admit that so far, I didn't understand the meaning of "touched" in the source and what the `tch` variable is doing. Could someone please explain the [logic behind that part of the PortAudio driver](https://github.com/hzulla/supercollider/blob/audiodriver/server/scsynth/SC_PortAudio.cpp#L262-L272)?

**Comment by hzulla** [comment]
Ok, [it's the input channel causing this](https://lists.columbia.edu/pipermail/portaudio/2016-April/000379.html). The moment you open an input channel, the callback function runs into buffer underruns by default even if there is no costly audio processing inside the ballback.

**Comment by brianlheim** [comment]
Could someone retest this on a more recent distro?

**Comment by patrickdupuis** [comment]
I'll try and test this in the coming week on Debian Testing and Fedora 25.

**Comment by patrickdupuis** [comment]
I've built SC with `-DAUDIOAPI=portaudio` on my Fedora 25 machine (which uses PulseAudio). 

When I start the server, the ALSA `default` device is chosen, which routes sound through PulseAudio. I am able to route sound into and out of SC via PA. I can also change the sample rate with `s.options.sampleRate`. The performance isn't reliable, however. I noticed 2 inexplicable server crashes.

When killing PulseAudio and trying the same test, the SC server boots choosing the ALSA device that corresponds to the actual hardware. The server did not crash like with PA, but it did kind of explode (usage goes above 100% with tons of dropouts) when I ran some heavy processing (http://sccode.org/1-4SB and http://sccode.org/1-4QL simultaneously).

Running SC with anything other than JACK is **experimental** and to be **use at your won risk**. SC with PortAudio as a back-end actually works, surprisingly. If I were trying to avoid JACK, I would go with ALSA direclty as the performance seemed slightly better and is probably easier to configure.

**Comment by b1tfl0w** [comment]
On my Raspberry pi with updated system and my laptop running Ubuntu 18.04 when I run Pure Data and Supercollider using Jack I can't get rid of a minimum latency and lots of glitch and crackle sound..while in Alsa is really smooth with no glitches, I can also play guitar with some effects without latency. I admit that this really surprised me.
So I came in to this post, followed instructions above and everything seem fine, audio is really good..with no audible latency! 
```
-> a ServerOptions
-> localhost
Booting server 'localhost' on address 127.0.0.1:57110.
Found 74 LADSPA plugins
Faust: supercollider.cpp: sc_api_version = 3
Faust: FaustJPverbRaw numControls=11
Faust: supercollider.cpp: sc_api_version = 3
Faust: FaustGreyholeRaw numControls=7
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:867:(find_matching_chmap) Found no matching channel map
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock

Device options:
  - ALSA : HDA Intel MID: ALC270 Analog (hw:0,0)   (device #0 with 2 ins 2 outs)
  - ALSA : HDA Intel MID: HDMI 0 (hw:0,3)   (device #1 with 0 ins 8 outs)
  - ALSA : Fast Track: USB Audio (hw:1,0)   (device #2 with 2 ins 2 outs)
  - ALSA : sysdefault   (device #3 with 128 ins 128 outs)
  - ALSA : front   (device #4 with 0 ins 2 outs)
  - ALSA : surround40   (device #5 with 0 ins 2 outs)
  - ALSA : surround51   (device #6 with 0 ins 2 outs)
  - ALSA : surround71   (device #7 with 0 ins 2 outs)
  - ALSA : hdmi   (device #8 with 0 ins 8 outs)
  - ALSA : pulse   (device #9 with 32 ins 32 outs)
  - ALSA : dmix   (device #10 with 0 ins 2 outs)
  - ALSA : default   (device #11 with 32 ins 32 outs)

Booting with:
  In: ALSA : Fast Track: USB Audio (hw:1,0)
  Out: ALSA : Fast Track: USB Audio (hw:1,0)
  Sample rate: 44100.000
  Latency (in/out): 0.009 / 0.009 sec
SC_AudioDriver: sample rate = 44100.000000, driver's block size = 64
SuperCollider 3 server ready.
Requested notification messages from server 'localhost'
localhost: server process's maxLogins (1) matches with my options.
localhost: keeping clientID (0) as confirmed by server process.
Shared memory server interface initialized
-> Synth('temp__0' : 1000)
-> an EventStreamPlayer
-> Synth('temp__1' : 1003)

```
